### PR TITLE
Correctly parse the Quorum config section

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Instead of incorrectly having an array of Quorums, there should only be one root Quorum with a set of nodes and sub-quorums.

### CODE REVIEW

**Review by Geod24 — CHANGES_REQUESTED** [review]
One thing that does not make sense to me:
In Yaml, if 2 entries are at the same indentation level, they are part of an array. The approach of special-casing `root` to mean it's one level higher looks like a hack. Why not just remove the `name` and reduce indentation by 1 level ?

**Comment by AndrejMitrovic** [comment]
How do you create sub-entries in YAML? I couldn't figure that out. That's why it was initially an array at the top level, with the first entry implicitly meaning "root".

**Comment by Geod24** [comment]
```yaml
quorum:
  # threshold as a percentage
  threshold: 66%
  nodes:
    - GBFDLGQQDDE2CAYVELVPXUXR572ZT5EOTMGJQBPTIHSLPEOEZYQQCEWN
    - GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5
  subQuorums:
    - name: quorum-1.1
      threshold: 66%
      nodes:
        - GBFDLGQQDDE2CAYVELVPXUXR572ZT5EOTMGJQBPTIHSLPEOEZYQQCEWN
        - GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5
      subQuorums:
        - name: quorum-1.1.1
           threshold: 66%
           nodes:
             - GBFDLGQQDDE2CAYVELVPXUXR572ZT5EOTMGJQBPTIHSLPEOEZYQQCEWN
             - GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5
             - GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5
```

Something like this ?

**Comment by AndrejMitrovic** [comment]
#209

**Comment by AndrejMitrovic** [comment]
The quorum config has been greatly simplified now, and I've added a unittest. This is the diff: https://github.com/bpfkorea/agora/compare/dde37dda787fecfb0fba8befaf95e394ad85e7f4..57326a31983a4c712b5b39384733751e164c9fba

**Comment by AndrejMitrovic** [comment]
Ah I have to fix the integration configs too. Edit: done.

File: source/agora/common/Config.d [hunk 1]
```diff
@@ -312,68 +308,130 @@ private BanManager.Config parseBanManagerConfig (Node node)
...
 {
-    struct PreQuorum
-    {
-        string threshold;
-        PublicKey[] nodes;
-        string[] sub_quorums;  // they're looked up later
-    }
+    import std.algorithm;
+    import std.exception;
+    enforce(level <= 3, "Cannot have more than 2 levels of sub-quorums.");
```
Comment (Geod24) [inline source/agora/common/Config.d:327]: Why?

File: source/agora/common/Config.d [hunk 2]
```diff
@@ -312,68 +308,130 @@ private BanManager.Config parseBanManagerConfig (Node node)
...
+              - threshold: 66%
+                nodes:
+                  - GBFDLGQQDDE2CAYVELVPXUXR572ZT5EOTMGJQBPTIHSLPEOEZYQQCEWN
+                  - GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5
+                  - GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5`;
+
+    auto node = Loader.fromString(conf_example).load();
+    auto quorum = parseQuorumSection(node["quorum"]);
+
+    auto expected = QuorumConfig(2,
```
Comment (Geod24) [inline source/agora/common/Config.d:375]: Beautiful

**Review by Geod24 — APPROVED** [review]
Very nice

**Comment by Geod24** [comment]
Needs a rebase, but ready for merge

**Comment by AndrejMitrovic** [comment]
Rebased.

File: source/agora/common/Config.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/common/Config.d:327]: I was basing it on Stellar. Maybe I should add it to the notes.

Actually funny thing, they increased the depth recently to enable 3 sub-quorums: https://github.com/stellar/stellar-core/pull/2233/commits/aa45d9eaa4ece2d2d6f9b2bcbdc5898ba5d18080

**Pull request merged**